File: density_estimation_MATLAB/test_background_scores.m
clear; close all; 

%===============================================================
% Data preparation
%  (Please copy and paste one section of the below.)
%===============================================================

X = load('background_scores.txt');
[m, ~] = size(X); 
randn('seed', 1); 
X = [mean(X(:,1:2), 2), X(:,3)] + 0.1*randn(m,2); 
B = X; 

figure;
scatter(X(:,1), X(:,2), '+'); title('statistics vs linear algebra'); 
xlabel('statistics score'); 
ylabel('linear algebra score'); 
saveas(gcf, 'stats_vs_linear', 'png'); 

input('press any key to continue ...\n'); 

figure; hist(B(:,1), 20); xlabel('statistics score'); 
saveas(gcf, 'stats_only', 'png'); 
figure; hist(B(:,2), 20); xlabel('linear algebra score'); 
saveas(gcf, 'linear_only', 'png'); 

input('press any key to continue ...\n'); 

%===============================================================
% Histogram for 2D data
%===============================================================

% Adjustable parameters:
bincount = 10;

mindata = min(B, [], 1);
maxdata = max(B, [], 1);
minlevel = floor(mindata);
maxlevel = ceil(maxdata);
leveldiff = (maxlevel - minlevel)/bincount;

bin = ceil(bsxfun(@rdivide, bsxfun(@minus, B, minlevel), leveldiff)); 
addpath('tensor_toolbox_2.6'); 
tmp = sptensor([bin, ones(m,1)], ones(m, 1), [bincount, bincount, m]); 
myhist3 = (bincount^2/prod(maxlevel - minlevel)) * sum(double(tmp), 3) ./ m; 

%% avoid for loops in matlab; 
% myhist3 = zeros(bincount,bincount,1);
% for i=1:size(B,1)
% 	bin = (B(i,:) - minlevel) ./ leveldiff;
% 	bin = ceil(bin);
% 	myhist3(bin(1),bin(2)) = myhist3(bin(1),bin(2)) + 1;
% end

figure; 
bar3(myhist3);
set(gca,'XTick',1:(bincount/10):bincount);
set(gca,'XTickLabel',minlevel(1):leveldiff(1)*(bincount/10):maxlevel(1));
set(gca,'YTick',1:(bincount/10):bincount);
set(gca,'YTickLabel',minlevel(2):leveldiff(2)*(bincount/10):maxlevel(2));
axis tight; 
xlabel('statistics score'); 
ylabel('linear algebra score'); 
saveas(gcf, 'histogram', 'png'); 

input('press any key to continue ...\n'); 

%===============================================================
% Kernel density estimation
%===============================================================

bincount = 100;
leveldiff = (maxlevel - minlevel)/bincount;

% Adjustable parameters:
kernelwidth = 0.60;

[gridx, gridy] = meshgrid( ...
    mindata(1) + leveldiff(1)*(1:bincount), ...
    mindata(2) + leveldiff(2)*(1:bincount) ...
    ); 

norm2B = sum(B.^2, 2); 
norm2grid = gridx(:).^2 + gridy(:).^2; 
dist2mat = bsxfun(@plus, norm2B, norm2grid') - 2 * B * [gridx(:), gridy(:)]'; 
kernelmatrix = exp(-dist2mat ./ (2 * kernelwidth^2)) ./ (kernelwidth^2 * (2*pi)); 

mykde2 = mean(kernelmatrix, 1); 

figure; 
% bar3(reshape(mykde2, [bincount, bincount]));
surf(reshape(mykde2, [bincount, bincount]));
set(gca,'XTick',1:(bincount/10):bincount);
set(gca,'XTickLabel',minlevel(1):leveldiff(1)*(bincount/10):maxlevel(1));
set(gca,'YTick',1:(bincount/10):bincount);
set(gca,'YTickLabel',minlevel(2):leveldiff(2)*(bincount/10):maxlevel(2));
xlabel('statistics score'); 
ylabel('linear algebra score'); 
title('kernel density estimator'); 
saveas(gcf, 'kde', 'png'); 

input('press any key to continue ...\n'); 

saveas(gcf, 'kde', 'png'); 

% overlay contour plot with density. 
figure; 
scatter(B(:,1), B(:,2), '+'); title('statistics vs linear algebra'); 
hold on; 
[c, h] = contour(gridx, gridy, reshape(mykde2, [bincount, bincount]));
xlabel('statistics score'); 
ylabel('linear algebra score'); 
saveas(gcf, 'overlay', 'png'); 

input('press any key to continue ...\n'); 

% fit with Gaussian cdf
mu = mean(B, 1); 
sigma = cov(B); 
tmppdf = mvnpdf([gridx(:),gridy(:)], mu, sigma);
figure; 
scatter(B(:,1), B(:,2), '+'); title('statistics vs linear algebra (Gaussian)'); 
hold on; 
[c, h] = contour(gridx, gridy, reshape(tmppdf, [bincount, bincount]));
xlabel('statistics score'); 
ylabel('linear algebra score'); 
saveas(gcf, 'gaussian_overlay', 'png'); 

figure; 
% bar3(reshape(mykde2, [bincount, bincount]));
surf(reshape(tmppdf, [bincount, bincount]));
set(gca,'XTick',1:(bincount/10):bincount);
set(gca,'XTickLabel',minlevel(1):leveldiff(1)*(bincount/10):maxlevel(1));
set(gca,'YTick',1:(bincount/10):bincount);
set(gca,'YTickLabel',minlevel(2):leveldiff(2)*(bincount/10):maxlevel(2));
xlabel('statistics score'); 
ylabel('linear algebra score'); 
title('Gaussian density'); 
saveas(gcf, 'Gaussian', 'png'); 

File: density_estimation_Python/test_background_scores.py
######################################################################
#The codes are based on Python2.7. 
#Please install numpy, scipy, matplotlib packages before using.
#Thank you for your suggestions!
#
#@version 1.0
#@author CSE6740/CS7641 TA
######################################################################
# ===============================================================
# Data preparation
# (Please copy and paste one section of the below.)
# ===============================================================
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
from scipy.stats import multivariate_normal

X = np.loadtxt('background_scores.txt')
m = X.shape[0]
np.random.seed(1)

mean01 = np.mean(X[:,0:2], axis = 1)
mean01 = mean01.reshape(mean01.shape[0], 1)
X = np.concatenate((mean01, X[:,2].reshape(X[:,2].shape[0], 1)), axis = 1) + \
     0.1 * np.random.randn(m, 2)
     
B = X 

plt.figure(1)
plt.scatter(X[:, 0], X[:, 1] , marker = '+') 
plt.title('statistics vs linear algebra')
plt.xlabel('statistics score')
plt.ylabel('linear algebra score')

#raw_input('press any key to continue ...\n')

plt.figure(2)
plt.hist(B[:, 0], 20)
plt.xlabel('statistics score')
plt.figure(3)
plt.hist(B[:, 1], 20)
plt.xlabel('linear algebra score')

#raw_input('press any key to continue ...\n')

# ===============================================================
# Histogram for 2D data
# ===============================================================

# Adjustable parameters:
bincount = 10

mindata = np.min(B, axis=0)
maxdata = np.max(B, axis=0)
minlevel = np.floor(mindata)
maxlevel = np.ceil(maxdata)
leveldiff = (maxlevel - minlevel) / bincount

myhist3 = np.zeros((bincount, bincount, 1))
for i in range(0, B.shape[0]):
    bin = (B[i,:] - minlevel) / leveldiff
    bin = np.ceil(bin)
    myhist3[bin[0]-1, bin[1]-1, 0] = myhist3[bin[0]-1, bin[1]-1, 0] + 1

fig = plt.figure(4)
ax = fig.add_subplot(111, projection='3d')

xpos = np.arange(myhist3.shape[0])
ypos = np.arange(myhist3.shape[1])
xposM, yposM = np.meshgrid(xpos, ypos)
zpos = myhist3
zpos = zpos.ravel()
dx = 0.5
dy = 0.5   
dz = zpos

ax.w_xaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_xaxis.set_ticklabels(np.arange(minlevel[1], maxlevel[1], leveldiff[1] * (bincount / 10)))
ax.w_yaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_yaxis.set_ticklabels(np.arange(minlevel[0], maxlevel[0], leveldiff[0] * (bincount / 10)))

values = np.linspace(0, 1, xposM.ravel().shape[0])
colors = cm.rainbow(values)
plt.xlabel('linear algebra score')
plt.ylabel('statistics score')

ax.bar3d(yposM.ravel(), xposM.ravel(), dz * 0, dx, dy, dz, color=colors) 

#raw_input('press any key to continue ...\n')

# ===============================================================
# Kernel density estimation
# ===============================================================

bincount = 100
leveldiff = (maxlevel - minlevel) / bincount

# Adjustable parameters:
kernelwidth = 0.60

gridx, gridy = np.meshgrid( mindata[0] + leveldiff[0] * np.arange(1,bincount + 1), \
           mindata[1] + leveldiff[1] * np.arange(1,bincount + 1), copy = False)

norm2B = np.sum(np.power(B,2), 1)
norm2grid = np.power(gridx.ravel(order='F'),2) + np.power(gridy.ravel(order='F'),2)

norm2B = norm2B.reshape(norm2B.shape[0], 1)
norm2grid = norm2grid.reshape(norm2grid.shape[0], 1)

gridxr = gridx.ravel(order='F')
gridyr = gridy.ravel(order='F')
gridxr = gridxr.reshape(gridxr.shape[0], 1)
gridyr = gridyr.reshape(gridyr.shape[0], 1)
dist2mat = norm2B + norm2grid.T - 2*B.dot(np.concatenate((gridxr, gridyr), axis = 1).T)

# Using Gaussian Kernel
kernelmatrix = np.exp(  -dist2mat/(2 * np.power(kernelwidth,2))  ) /   \
                (np.power(kernelwidth, 2) * (2 * np.pi))

mykde2 = np.mean(kernelmatrix, axis = 0)

fig = plt.figure(5)
ax = fig.add_subplot(111, projection='3d')

X1 = np.arange(bincount)
Y = np.arange(bincount)
X1, Y = np.meshgrid(X1, Y, copy = False)
Z = mykde2.reshape(bincount, bincount).T
ax.plot_surface(X1, Y, Z, cmap='coolwarm', rstride=2, cstride=2)

ax.w_xaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_xaxis.set_ticklabels(np.arange(minlevel[1], maxlevel[1], leveldiff[1] * (bincount / 10)))
ax.w_yaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_yaxis.set_ticklabels(np.arange(minlevel[0], maxlevel[0], leveldiff[0] * (bincount / 10)))   
plt.gca().invert_xaxis()
plt.xlabel('linear algebra score')
plt.ylabel('statistics score')
plt.title('kernel density estimator')

#raw_input('press any key to continue ...\n')

# overlay contour plot with density. 
plt.figure(6)
plt.scatter(X[:, 0], X[:, 1] , marker = '+') 
plt.hold(True)
plt.contour(gridx, gridy, mykde2.reshape(bincount, bincount))
plt.title('statistics vs linear algebra')
plt.xlabel('statistics score')
plt.ylabel('linear algebra score')

#raw_input('press any key to continue ...\n')

# fit with Gaussian cdf
mu = np.mean(B, 0)
sigma = np.cov(B.T)

tmppdf = multivariate_normal.pdf(np.concatenate((gridxr, gridyr), axis = 1), mean = mu, cov = sigma)
plt.figure(7)
plt.scatter(B[:, 0], B[:, 1] , marker = '+')
plt.title('statistics vs linear algebra (Gaussian)')
plt.hold(True)
plt.contour(gridx, gridy, tmppdf.reshape(bincount, bincount))
plt.xlabel('statistics score')
plt.ylabel('linear algebra score')

fig = plt.figure(8)
ax = fig.add_subplot(111, projection='3d')
X1 = np.arange(bincount)
Y = np.arange(bincount)
X1, Y = np.meshgrid(X1, Y)
Z = tmppdf.reshape(bincount, bincount).T
ax.plot_surface(X1, Y, Z, cmap='coolwarm', rstride=2, cstride=2)

ax.w_xaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_xaxis.set_ticklabels(np.arange(minlevel[1], maxlevel[1], leveldiff[1] * (bincount / 10)))   
ax.w_yaxis.set_ticks(np.arange(1, bincount, bincount / 10))
ax.w_yaxis.set_ticklabels(np.arange(minlevel[0], maxlevel[0], leveldiff[0] * (bincount / 10)))

plt.gca().invert_xaxis()
plt.xlabel('linear algebra score')
plt.ylabel('statistics score')
plt.title('Gaussian density')
plt.show()